proposals/ssss-distributed-actor-runtime.md

```diff
@@ -0,0 +1,1798 @@
+  /// the 'distributed actor' will gain an automatically synthesized conformance to 'Codable' as well.
+  associatedtype ActorID: Sendable & Hashable
+  
+  /// The specific type of the invocation encoder that will be created and populated 
+  /// with details about the invocation when a remote call is about to be made.
+  ///
+  /// The populated instance will be passed to the remoteCall from where it can be
+  /// used to serialize into a message format in order to perform the remote invocation.
+  associatedtype InvocationEncoder: DistributedTargetInvocationEncoder
+  
+  /// The specific type of invocation decoder used by this actor system.
+  ///
+  /// An instance of this type must be passed to `executeDistributedTarget` which
+  /// extracts arguments and applies them to the local target of the invocation.
+  associatedtype InvocationDecoder: DistributedTargetInvocationDecoder
+
+  /// The serialization requirement that will be applied to all distributed targets used with this system. 
+  ///
+  /// An actor system is still allowed to throw serialization errors if a specific value passed to a distributed
+  /// func violates some other restrictions that can only be checked at runtime, e.g. checking specific types
+  /// against an "allow-list" or similar. The primary purpose of the serialization requirement is to provide
+  /// compile time hints to developers, that they must carefully consider evolution and serialization of 
+  /// values passed to and from distributed methods and computed properties.
+  associatedtype SerializationRequirement
+    where SerializationRequirement == InvocationEncoder.SerializationRequirement,
+          SerializationRequirement == InvocationDecoder.SerializationRequirement
+  
+  // ==== ---------------------------------------------------------------------
+  // - MARK: Actor Lifecycle
+
+  /// Called by a distributed when it begins its initialization (in a non-delegating init).
+  /// 
+  /// The returned 'ID' stored by the distributed actor and is used to uniquely identify and 
+  /// locate the actor within the system. Once 'actorReady' is called resolving this 'ID'
+  /// with 'resolve(_:as:)' should return the same instance was just assigned this identity.
+  ///
+  /// The system should take special care to not assign two actors the same 'ID', and the 'ID'
+  /// must remain valid until it is resigned (see 'resignID(_:)').
+  func assignID<Act>(_ actorType: Act.Type) -> ActorID
+      where Act: DistributedActor, 
+            Act.ID == ActorID
+
+  /// Called when the distributed actor has been fully initialized (meaning that all of its,
+  /// stored properties have been assigned an initial value).
+  ///
+  /// The distributed actor will pass its 'self' to 'actorReady(_:)' as it is fully initialized,
+  /// and from that moment onwards it must be possible to resolve it using the 'resolve(_:as:)'
+  /// method on the system.
```

Comment:
I reworded it a lot since we simplified the semantics based on your feedback :)

```
  /// Automatically called by in every distributed actor's non-delegating initializer.
  ///
  /// The call is made specifically before the `self` of such distributed actor is about to
  /// escape, e.g. via a function call, closure or otherwise. If no such event occurs the
  /// call is made at the end of the initializer.
  ///
  /// The passed `actor` is the `self` of the initialized actor, and its `actor.id` is expected
  /// to be of the same value that was assigned to it in `assignID`.
  /// 
  /// After the ready call returns, it must be possible to resolve it using the 'resolve(_:as:)' 
  /// method on the system.
```